src/libstd/thread/mod.rs

```diff
@@ -1060,6 +1060,34 @@ impl Thread {
     fn cname(&self) -> Option<&CStr> {
         self.inner.name.as_ref().map(|s| &**s)
     }
+
+    /// Change the name of a thread
+    ///
+    /// ```
+    /// use std::thread;
+    ///
+    /// let builder = thread::Builder::new()
+    ///     .name("foo".into());
+    ///
+    /// let handler = builder.spawn(|| {
+    ///     assert_eq!(thread::current().name(), Some("foo"));
+    ///     thread::current().set_name("bar");
+    ///     assert_eq!(thread::current().name(), Some("bar"));
+    /// }).unwrap();
+    ///
+    /// handler.join().unwrap();
+    /// ```
+    #[unstable(feature = "thread_rename", reason = "the compiler threw an error at me ;)",
+               issue = "27802")]
+    pub fn set_name(&mut self, name: &str) {
+        let cname = CString::new(name).expect("thread name may not contain interior null bytes");
+        imp::Thread::set_name(&cname);
+        let target: *const Option<CString> = &self.inner.name;
+        unsafe {
+            let target = ::mem::transmute::<*const Option<CString>, &mut Option<CString>>(target);
+            *target = Some(cname);
```

Comment:
Mutating through a shared pointer like this is illegal, inside and outside of `unsafe` blocks.

For the inner thread we have an `Arc<Inner>` and no *interior mutability*, in particular no interior mutability around the cname field, so it's completely impossible to mutate that field, unless we get mutable access to it like with `Arc::get_mut`, but that method is only useful when there is no other handle to the same Arc.

So to make it possible to change cname, interior mutability is needed. Usually this means `Mutex`.

Unfortunately. The method `pub fn name(&self) -> Option<&str>` just a bit up in the file will exclude any possibility of using interior mutability. This is where even more creativity needs to pour in to solve it. :smile: